src/serial.rs

```diff
@@ -505,39 +515,40 @@ macro_rules! usart {
                     $(, $synchronous: bool)?
                 ) -> Result<Self, config::InvalidConfig>
                 {
-                    use crate::stm32::usart1::cr2::STOP_A as STOP;
-                    use self::config::*;
+                    // Enable clock for USART and reset
+                    prec.enable().reset();
 
+                    let ker_ck = Self::kernel_clk_unwrap(clocks);
+                    let mut serial = Serial { usart, ker_ck };
                     let config = config.into();
+                    serial.usart.cr1.reset();
+                    serial.configure(&config $(, $synchronous )?);
 
-                    // Enable clock for USART and reset
-                    prec.enable().reset();
+                    Ok(serial)
+                }
 
-                    // Get kernel clock
-	                let usart_ker_ck = Self::kernel_clk_unwrap(clocks).0;
+                fn configure(&mut self, config: &config::Config $(, $synchronous: bool)?) {
```

Comment:
This is an internal method, but could you add a comment as a reminder that it can only be called when the peripheral is disabled?